File: contracts/fakeout.js
const http = require('http');
const url = require('url');
const fs = require('fs');
var crypto  = require('crypto');

function generateBody(jsonSchema) {
  if(jsonSchema.default) return jsonSchema.default;
  switch(jsonSchema.type) {
    case 'array':
      const numItems = Math.floor(Math.random() * 10) + 1;
      let items = [];
      for(let i = 0; i < numItems; i++) {
        items.push(generateBody(jsonSchema.items));
      }
      return items;
    case 'string':
      const buf = crypto.randomBytes(8);
      return buf.toString('hex');
    case 'number':
      return Math.random();
    case 'object':
      return Object.keys(jsonSchema.properties).reduce((h, key) => {
        h[key] = generateBody(jsonSchema.properties[key]);
        return h;
      }, {});
    default:
      return '';
  }
}

const server = http.createServer((request, response) => {
  const path = url.parse(request.url).pathname;
  const schemaFilePath = `schemas/${path}/${request.method}.contract.json`;
  const schema = JSON.parse(fs.readFileSync(schemaFilePath));

  response.setHeader('Content-Type', 'application/json');
  response.end(JSON.stringify(generateBody(schema.response.bodySchema)));
});

var port = 8081;
server.listen(port, () => {
  console.log(`Good job, you booted on port: ${port}`);
});

File: contracts/fakeout.rb
require 'json'
require 'securerandom'
require 'rack'

def generate_body(json_schema)
  return json_schema["default"] if json_schema["default"]

  case json_schema["type"]
    when "array"
      rand(1..10).times.map do
        generate_body(json_schema["items"])
      end
    when "string"
      SecureRandom.hex(8)
    when "number"
      rand
    when "object"
      json_schema["properties"].each_with_object({}) do |(k, v), object|
        object[k] = generate_body(v)
      end
    else
      ""
  end
end

class App
  def call(env)
    req = Rack::Request.new(env)
    puts "Responding to: #{req.request_method} #{req.path}"
    schema_path = File.join("schemas", req.path, "#{req.request_method}.contract.json")
    if File.exists?(schema_path)
      schema = JSON.parse(File.read(schema_path))

      [
        schema['response']['status'],
        {'Content-Type' => 'application/json'},
        [JSON.dump(generate_body(schema['response']['bodySchema']))]
      ]
    else
      [
        404,
        {'Content-Type' => 'application/json'},
        [JSON.dump({errors: ["schema not found for endpoint #{req.request_method} #{req.path}"]})]
      ]
    end
  end
end

app = Rack::Builder.new { run App.new }
Rack::Server.start app: app, Port: 8081
